### PR TITLE
[FLINK-33985][runtime] Support obtain all partitions existing in cluster through ShuffleMaster.

### PR DESCRIPTION
**Pull request by JunRuiLee**
## What is the purpose of the change

[FLINK-33985][runtime] Support obtain all partitions existing in cluster through ShuffleMaster.

## Brief change log

  - Support obtain all partitions existing in cluster through ShuffleMaster.
  - TaskExecutor will not release partitions immediately when JM Failover enabled.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Review by zhuzhurk — COMMENTED** [review]
The change looks generally good to me.
Just several minor comments.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java [hunk 1]
```diff
@@ -1900,6 +1906,74 @@ void testJobMasterAcceptsSlotsWhenJobIsRestarting() throws Exception {
         }
     }
 
+    @Test
+    void testGetAllPartitionWithMetrics() throws Exception {
```
[the remark this answers is not in the document]
Comment (JunRuiLee) [reply flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java:1910]: Currently, the changes in `NettyShuffleMaster` haven't been introduced into the production code path, and they will be covered by the upcoming PR. Therefore, I'd prefer to include the test for `NettyShuffleMaster#getAllPartitionWithMetrics()` in the subsequent PR

**Comment by JunRuiLee** [comment]
@zhuzhurk Thanks for your review, I've updated this pr accordingly, PTAL.

File: flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/NettyShuffleMaster.java [hunk 2]
```diff
@@ -62,7 +61,7 @@ public class NettyShuffleMaster implements ShuffleMaster<NettyShuffleDescriptor>
 
     @Nullable private final TieredInternalShuffleMaster tieredInternalShuffleMaster;
 
-    private final Map<JobID, JobMasterGateway> jobMasters = new HashMap<>();
+    private final Map<JobID, JobShuffleContext> jobMasterShuffleMasters = new HashMap<>();
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/NettyShuffleMaster.java:64]: I prefer to name it as `jobShuffleContexts`

**Comment by JunRuiLee** [comment]
@zhuzhurk Fixed, PTAL~

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Pull request closed without merging**